### PR TITLE
HL-153

### PR DESCRIPTION
**Pull request by bazmatic**
https://dltx.atlassian.net/browse/HL-153

* Added `MarketCollateralised`: A market that only requests the smallest amount of cover necessary to cover bets, assuming that each market can only have one winning proposition. It only ever sends the bet amounts to the vault. It never explicitly returns amounts sent from the vault. **It keeps its own supply of collateral.**
* Added `MarketWithoutProtection`: A market that does not adjust odds at all. For testing only.
* Added preset `MarketCollateralisedWithoutProtection`: A combination of each of the above - collateralised, no odds adjustment

MarketCollateralised
This market tracks the "most expensive" possible proposition for each market. 

It ensures that there is sufficient "cover" to pay out this proposition. If another proposition becomes more "expensive", it obtains extra cover for the new exposure.

To do this, it checks to see if there is any leftover cover in the Market contract first, and then requests the difference from the Vault.

Notes:
* Some `private` properties in `Market.sol` have been changed to internal that they are accessible to descendant contracts (eg MarketCollateralised).
* Ran code reformatter on contract code, so there are some formatting changes.
* Moved test functions to own unit

### CODE REVIEW

File: contracts/Market.sol [hunk 1]
```diff
@@ -110,8 +118,8 @@ contract Market is IMarket, Ownable, ERC721 {
 		return _getExpiry(index);
 	}
 
-	function getMarketTotal(bytes16 marketId) external view returns (uint256) {
-		return _marketTotal[marketId];
+	function getMarketTotalWagers(bytes16 marketId) external view returns (uint256) {
```
Comment (bitcoinbrisbane) [inline contracts/Market.sol:121]: Not sure i like the name.  Maybe totalValueLocked?

File: contracts/Market.sol [hunk 2]
```diff
@@ -36,14 +38,20 @@ contract Market is IMarket, Ownable, ERC721 {
...
 
-	// PropositionID => amount bet
+	// PropositionID => amount that could be paid out for this proposition
 	mapping(bytes16 => uint256) private _potentialPayout;
 
+	// MarketID => current maximum payout for the market
+	mapping(bytes16 => uint256) private _maxMarketPayout;
+
+	// MarketID => PropositionID => proposition with the maximum payout for this market
+	//mapping(bytes16 => bytes16) private _maxMarketPayoutProposition;
```
Comment (OptimusOpus) [inline contracts/Market.sol:53]: Are we using this mapping? Should we remove it if we are not using it?

**Review by OptimusOpus — APPROVED** [review]
Just one comment and sonarcloud isn't happy, but LGTM

File: contracts/Market.sol (hunk 1 above)
Comment (bazmatic) [reply contracts/Market.sol:122]: It represents the total amount of wager amounts that have been placed for the given market. It used to be `getMarketTotal` but I decided to make it more explicit. `totalValueLocked` sounds ambiguous to me. Why would you prefer this term?

File: contracts/Market.sol [hunk 3]
```diff
@@ -172,34 +189,28 @@ contract Market is IMarket, Ownable, ERC721 {
...
 
+		// If the pool is not sufficient to cover a new bet
+		if (pool == 0) return 1;
 		// exclude the current total potential payout from the pool
 		if (_potentialPayout[propositionId] > pool) {
 			return 1;
 		}
 
-		pool -= _potentialPayout[propositionId]; 
+		pool -= _potentialPayout[propositionId]; //TODO: Should be _totalExposure;
```
Comment (brettvniekerk) [inline contracts/Market.sol:201]: should this TODO be completed as part of this PR?

File: contracts/MarketGreedy.sol [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    // MarketID => amount of cover taken for this market
+	mapping(bytes16 => uint256) private _marketCover;
+	uint256 internal _totalCover;
+	mapping(bytes16 => bytes16) internal _mostExpensivePropositionId;
+
+    function getMarketCover(bytes16 marketId) external view returns (uint256) {
+		return _marketCover[marketId];
+	}
+
+	//TODO: Should return the dedection to total exposure to be
```
Comment (brettvniekerk) [inline contracts/MarketGreedy.sol:16]: same here

**Review by brettvniekerk — APPROVED** [review]
LGTM just checking the TODO's

File: hardhat.config.ts [hunk 5]
```diff
@@ -35,6 +35,7 @@ export default {
 		hardhat: {
 			chainId: 1337,
 			saveDeployment: true,
+			allowUnlimitedContractSize: true,
```
Comment (bitcoinbrisbane) [inline hardhat.config.ts:38]: Why do we have this?  If there's a problem, we have a problem.

File: contracts/MarketGreedy.sol [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+	function getTotalCover() external view returns (uint256) {
+		return _totalCover;
+	}
+
+	// Allow the Vault to reclaim any cover it has provided, provided it is not currently covering any bets
+	function reclaimCover() external onlyOwner {
+		if (_totalCover > _totalExposure) {
+			IERC20(_vault.asset()).transfer(address(_vault), _totalCover - _totalExposure);
+			_totalCover = _totalExposure;
```
Comment (bitcoinbrisbane) [inline contracts/MarketGreedy.sol:86]: Should be the other way around "checks effects interactions"

File: contracts/MarketGreedy.sol [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+		}
+
+		//If paying out the most expensive proposition,
+		if (isMostExpensiveProposition(_bets[index].propositionId, _bets[index].marketId)) {
+			// Deduct from total exposure
+			_totalExposure -= (_bets[index].payout - _bets[index].amount);
+		}
+	}
+
+	function isMostExpensiveProposition(bytes16 propositionId, bytes16 marketId) internal view returns (bool) {
```
Comment (bitcoinbrisbane) [inline contracts/MarketGreedy.sol:41]:
```suggestion
	function _isMostExpensiveProposition(bytes16 propositionId, bytes16 marketId) internal view returns (bool) {
```

File: contracts/MarketGreedy.sol [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+			// Send the payout to the bet owner
+			IERC20(underlying).transfer(_bets[index].owner, _bets[index].payout);
+			_totalCover -= _bets[index].payout - _bets[index].amount;  
+		} else {
+			// Send the bet amount to the vault
+			IERC20(underlying).transfer(address(_vault), _bets[index].amount);	
+		}
+
+		//If paying out the most expensive proposition,
+		if (isMostExpensiveProposition(_bets[index].propositionId, _bets[index].marketId)) {
```
Comment (bitcoinbrisbane) [inline contracts/MarketGreedy.sol:35]:
```suggestion
		if (_isMostExpensiveProposition(_bets[index].propositionId, _bets[index].marketId)) {
```

Comment (bitcoinbrisbane) [reply contracts/MarketGreedy.sol:35]: We should standardise here.  In other spots we have the syntax if function == true

File: contracts/MarketGreedy.sol [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+	//TODO: Should return the dedection to total exposure to be 
+	function _payout(uint256 index, bool result) internal override {
+		require(
+			_bets[index].payoutDate < block.timestamp,
+			"_settle: Payout date not reached"
+		);
+		address underlying = _vault.asset();
+
+		//_totalInPlay -= _bets[index].amount;
```
Comment (bitcoinbrisbane) [inline contracts/MarketGreedy.sol:24]: Remove this guy

File: contracts/MarketGreedy.sol [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+    function getMarketCover(bytes16 marketId) external view returns (uint256) {
+		return _marketCover[marketId];
+	}
+
+	//TODO: Should return the dedection to total exposure to be 
+	function _payout(uint256 index, bool result) internal override {
+		require(
+			_bets[index].payoutDate < block.timestamp,
+			"_settle: Payout date not reached"
```
Comment (bitcoinbrisbane) [inline contracts/MarketGreedy.sol:20]:
```suggestion
			"_payout: Payout date not reached"
```

File: contracts/MarketGreedy.sol [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+	mapping(bytes16 => bytes16) internal _mostExpensivePropositionId;
+
+    function getMarketCover(bytes16 marketId) external view returns (uint256) {
+		return _marketCover[marketId];
+	}
+
+	//TODO: Should return the dedection to total exposure to be 
+	function _payout(uint256 index, bool result) internal override {
+		require(
+			_bets[index].payoutDate < block.timestamp,
```
Comment (bitcoinbrisbane) [inline contracts/MarketGreedy.sol:19]: I don't think this is correct, the payout date should be greater than now.

File: contracts/MarketGreedy.sol [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+			"_settle: Payout date not reached"
+		);
+		address underlying = _vault.asset();
+
+		//_totalInPlay -= _bets[index].amount;
+		if (result == true) {
+			// Send the payout to the bet owner
+			IERC20(underlying).transfer(_bets[index].owner, _bets[index].payout);
+			_totalCover -= _bets[index].payout - _bets[index].amount;
```
Comment (bitcoinbrisbane) [inline contracts/MarketGreedy.sol:28]: Same here.

```suggestion
                       _totalCover -= _bets[index].payout - _bets[index].amount; 
			IERC20(underlying).transfer(_bets[index].owner, _bets[index].payout);
			 
```

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
Few things to address

File: contracts/MarketGreedy.sol [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+                result = amountToTransfer;
+				_totalCover += result;
+            } else {
+				result = amountRequired;
+			}
+            _marketCover[marketId] += amountRequired;     
+		}
+		
+		//console.log("_totalCover now: %s", _totalCover);
+		//console.log("_obtainCover() end");
```
Comment (OptimusOpus) [inline contracts/MarketGreedy.sol:74]: 2 sneaky `console.log()`s here we want to remove

**Review by OptimusOpus — DISMISSED** [review]
Just one comment

**Comment by bitcoinbrisbane** [comment]
Should move the function getTotalCover to the top with the other properties.

File: contracts/MarketGreedy.sol [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+	function isMostExpensiveProposition(bytes16 propositionId, bytes16 marketId) internal view returns (bool) {
+		return keccak256(abi.encodePacked(propositionId)) == keccak256(abi.encodePacked(_mostExpensivePropositionId[marketId]));
+	}
+
+	// Overridden to make this "greedy" - it will hold on to the cover amounts for future bets
+    // If the payout for this proposition will be greater than the amount of cover set aside for the market
+	// Return the new exposure amount
+	function _obtainCover(bytes16 marketId, bytes16 propositionId, uint256 wager, uint256 payout) internal override returns (uint256) {
+		uint256 result = 0;
```
Comment (bitcoinbrisbane) [inline contracts/MarketGreedy.sol:49]: Can save gas by doing this...

```suggestion
		uint256 result;
```

File: contracts/MarketGreedy.sol [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+		//console.log("_obtainCover() end");
+		return result;
+	}
+
+	function getTotalCover() external view returns (uint256) {
+		return _totalCover;
+	}
+
+	// Allow the Vault to reclaim any cover it has provided, provided it is not currently covering any bets
+	function reclaimCover() external onlyOwner {
```
Comment (bitcoinbrisbane) [inline contracts/MarketGreedy.sol:83]: I think this should be called `withdraw`

File: contracts/MarketGreedy.sol (hunk 4 above)
Comment (bazmatic) [reply contracts/MarketGreedy.sol]: Removed comment

File: contracts/Market.sol (hunk 3 above)
Comment (bazmatic) [reply contracts/Market.sol:201]: I think this is a fine point that needs to be discussed separately.

File: contracts/MarketGreedy.sol (hunk 11 above)
Comment (bazmatic) [reply contracts/MarketGreedy.sol:18]: No. This would mean than payouts could only happen before the payout date and fail afterwards.

**Review by OptimusOpus — APPROVED** [review]
LGTM

File: contracts/MarketGreedy.sol (hunk 11 above)
Comment (brettvniekerk) [reply contracts/MarketGreedy.sol:18]: Yeah it looks like the `_payout` function should only be called _after_ the date has been reached

**Review by brettvniekerk — APPROVED** [review]

**Review by bitcoinbrisbane — APPROVED** [review]

**Review by OptimusOpus — APPROVED** [review]

**Review by bitcoinbrisbane — APPROVED** [review]

**Pull request merged**